### DISCUSSION TITLE
3.12.3 introduces a breaking change when using DjangoRedis 4.12.1

### DISCUSSION THREAD

**Discussion by Bradleywboggs** [encode/django-rest-framework#7870] [Potential Issue]
3.12.3 introduces a breaking change when using DjangoRedis 4.12.1 

For a request without a cached throttling history, the change in the class `SimpleRateThrottle`,
at `rest_framework/throttling.py::124`: 
`self.history = self.cache.get(self.key, deque())`
causes a `TypeError` to be raised in the same class in the `throttle_success` method specifically at line 141: 
` self.cache.set(self.key, self.history, self.duration`

The root of the problem is that DjangoRedis json-serializes the value before inserting into the cache, but `deques` are not json serializable. So the solution would appear to be to cast the value as a deque on `get` and  as a list on`set` or just  keep the history as a list.
e.g.

```
self.history = deque(self.cache.get(self.key, []))
....
self.cache.set(self.key, list(self.history), self.duration)
```

@tomchristie

**Comment by lovelydinosaur** [comment]
Good example of how even innocuous looking changes have an attached risk.

---

🙂 - "Hey, PR #7849 for keeping the history as a deque instead of list. Makes sense, right?"
  "...Um, I guess, yup. There might be some risks we've not foreseen." - 🤔
😀 - "Can't see any."
  "No me neither. Okay, I guess this may as well go in." - 👍

*... (not much) time passes ...*

**\<unforeseen risks become surfaced\>** 😑

---

Anyways, just another data point for erring on the side of keeping code churn to an absolute minimum wherever possible.

I'll go ahead and revert this, and likely roll a new release.

**Reply by Bradleywboggs** [reply]
Haha! I totally get it! Thanks for the quick response.
